### ISSUE TITLE
Dead letter facility for event-trigger webhooks

### ISSUE THREAD

**Issue by mike-thompson-day8** [hasura/graphql-engine#2097]
We'd like a mechanism which "lets us know" that webhooks (for event-triggers) are failing to run (after the `max_retries` and `retry_interval` based process is exhausted).  

Associated with an installation should be some kind of "dead-letter queue" for these event-triggers which have failed to successfully webhook. 

How to implement it? 
  - Perhaps it is just a query we can schedule to run which returns all failed event-triggers?  
  - Perhaps this dead-letter-queue should itself be a webhook to which failed events can be POSTed. Up to me then to ensure the dead-letter webhook works by emailing or slack-notifying the problem

**Comment by mike-thompson-day8** [comment]
So, just to be clear: for webhooks to be robust, this is an essential idea.  `event-triggers` are an important part of our architecture, and it is "a bad thing" that they can potentially be failing silently. 

So, I wouldn't be labelling this issue as "an idea".  I'd say it is a bug that this mechanism isn't there already.  :-)

**Comment by coco98** [comment]
@tirumaraiselvan and I'd been discussing error handling for dead events too so this is definitely on point. 

@tirumaraiselvan @dsandip Let's take this up quickly!

**Comment by tirumaraiselvan** [comment]
@mike-thompson-day8 Yes, something like this has been in the plan.

For the time being, you may use the `delivery_info` in the payload with which you can determine if this is the last retry or not and maybe do some custom error handling in the webhook itself.

**Comment by tirumaraiselvan** [comment]
@mike-thompson-day8  Some more thoughts on this:

How do you intend to use the dead letter queue? As a notification system to check for failures or as another retry mechanism for guaranteed ack? We must keep in mind that we cannot expect 100% guaranteed ack from any dead letter endpoint as well. Using a dead letter queue as a tool for guaranteed processing might only cause more obfuscation.

So, a dead letter endpoint might be useful for notification use-cases which tells you to run diagnostics using your custom playbook. We can give all APIs to fetch dead events, remove events and so on which you use in your playbook. Again, the notification is not 100% guaranteed so it should be part of one of many alarms in your system.

**Comment by mike-thompson-day8** [comment]
If I were to misconfigure an event trigger by mistyping the webhook URL.  Instead of `http://.....`  I put `htt://` or something equally bad.  

At the moment, the `max_retries` and `retry_interval` won't help to deliver an event.  It will ultimately completely fail.  But it will currently fail silently.  We'll never know there's a problem. 

I'm looking for a way to be told there's a problem via a channel like slack or email. 

BTW, after I fix the problem, I'd probably be wishing that there was a way to reprocess all failed triggers.  But that's another matter.

**Comment by tirumaraiselvan** [comment]
Cool. So its more of a notification system. In that case, are you okay with the "fire and forget" design of the dead letter endpoint? Whenever we encounter a failed event, we hit the notification endpoint once.

**Comment by tirumaraiselvan** [comment]
@mike-thompson-day8 We are beginning to work on this issue. Since a "fire and forget" notification would not work for reliability, we are thinking of providing a "monitoring" webhook instead. The monitoring webhook will periodically receive stats about the event trigger including number of pending events/failed events, breakdown by the hour etc?

Since the monitoring webhook is running forever, you are effectively guaranteed to recv the notification sometime.

**Comment by mike-thompson-day8** [comment]
Yeah, I think that would work.  

This monitoring endpoint could be a serverless endpoint somewhere that sends an email in the case it sees many failed webhooks or many being retried.  Or something.  This satisfied my goal of "not failing silently". 

So, questions:
  1. would it be one monitoring webhook per regular webhook?  
  2. a configurable cadence? Once every hour? Once every 5 mins? 

Aside:  I'll mention again the related issue of "being able to request that all failures be retried again" (usecase: after some configuration problem is fixed)

**Comment by taylorhakes** [comment]
I would like to add another issue. What if it’s a network issue on the sender side? I think these events need to be written to a table primarily. We can have the notify webhook/endpoint, but dead letter events should be written to a table 1st, so they aren’t dropped in the case of a network issue.


I haven’t looked too much how events are implemented, but do the events get lost if the server restarts while processing an event? If so, these 2 problems are closely related and should be able to be fixed with the same mechanism.

**Comment by tirumaraiselvan** [comment]
@taylorhakes No, none of the events are lost as they are persisted for reliability.

**Comment by fpieper** [comment]
@tirumaraiselvan what is the current state of this issue? Because currently the possibility of silent failure keeps me off of using event triggers. Thanks a lot in advance :)

**Comment by dionjwa** [comment]
It's been a while and I'd like to know the status of this issue. We've hit this problem in our production systems. Is there any unofficial query to make that would alleviate the problem for now?

**Comment by tirumaraiselvan** [comment]
@dionjwa We are scoping this feature out this quarter. It will be great to get your input on this. If you are interested, we can setup a time through email. My id is `tiru@hasura.io`

**Comment by wawhal** [comment]
I agree with @mike-thompson-day8's "Better" solution i.e to have a webhook that Hasura calls while marking an event as complete (failed or succeeded). This way the users can capture the dead events, do their business logic and update their database if needed.

**Comment by johnkoehn** [comment]
Any interest in doing this still? Would be a 🔥 feature
